eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1327 @@
+
+  g_set_error (error,
+               EKNR_ERROR,
+               EKNR_ERROR_UNKNOWN_LEGACY_SOURCE,
+               "HTML is not server templated, but no renderer exists for %s",
+               source);
+
+  return NULL;
+}
+
+static GVariant *
+get_wrapper_css_files (const char * const *custom_css_files)
+{
+  GPtrArray *wrapper_css_files = g_ptr_array_new ();
+  const char *clipboard_css = "clipboard.css";
+  const char *share_actions_css = "share-actions.css";
+
+  g_ptr_array_add (wrapper_css_files, (gpointer) clipboard_css);
+  g_ptr_array_add (wrapper_css_files, (gpointer) share_actions_css);
+
+  for (; *custom_css_files != NULL; ++custom_css_files)
+    g_ptr_array_add (wrapper_css_files, (gpointer) *custom_css_files);
+
+  g_ptr_array_add (wrapper_css_files, NULL);
+
+  return g_variant_new_strv ((const char * const *) wrapper_css_files->pdata,
+                             -1);
+}
+
+static GVariant *
+get_wrapper_js_files (void)
+{
+  const char * const wrapper_js_files[] = {
+    "jquery-min.js",
+    "clipboard-manager.js",
+    "crosslink.js",
+    "chunk.js",
+    "share-actions.js",
+    NULL
+  };
+
+  return g_variant_new_strv (wrapper_js_files, -1);
+}
+
+static GVariant *
+get_wrapper_outgoing_links_json (const char * const                *outgoing_links,
+                                 EknrOutgoingLinkDeterminationFunc  determination_func,
+                                 gpointer                           user_data)
+{
+  g_autoptr(JsonBuilder) builder = json_builder_new ();
+  g_autoptr(JsonNode) node = NULL;
+  g_autoptr(JsonGenerator) generator = NULL;
+  g_autofree char *json_string = NULL;
+
+  json_builder_begin_array (builder);
+
+  for (; *outgoing_links != NULL; ++outgoing_links)
+    {
+      const char *link = *outgoing_links;
+      g_autofree char *internal_link = (*determination_func) (link, user_data);
+
+      if (internal_link)
+        json_builder_add_string_value (builder, internal_link);
+    }
+
+  json_builder_end_array (builder);
+  node = json_builder_get_root (builder);
+  generator = json_generator_new ();
+
+  json_generator_set_root (generator, node);
+  json_string = json_generator_to_data (generator, NULL);
+
+  return g_variant_new_string (json_string);
+}
+
+static GPtrArray *
+get_wrapper_parent_featured_sets (const char * const *tags,
+                                  GHashTable          *set_map)
+{
+  g_autoptr(GPtrArray) parent_featured_sets = g_ptr_array_new_with_free_func ((GDestroyNotify) eknr_boxed_set_info_unref);
+
+  for (; *tags != NULL; ++tags)
+    {
+      const char      *tag = *tags;
+      EknrBoxedSetInfo *set = NULL;
+
+      if (strncmp (tag, "Ekn", 3) == 0)
+        continue;
+
+      set = g_hash_table_lookup (set_map, tag);
+
+      if (set == NULL)
+        continue;
+
+      if (!set->featured)
+        continue;
+
+      g_ptr_array_add (parent_featured_sets,
+                       eknr_boxed_set_info_ref (set));
```

Comment:
As mentioned elsewhere, there is no need to ref/unref these, as `set_map` owns the sets and will always stay alive.